src/nupic/support/group_by.py

```diff
@@ -0,0 +1,161 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+
+
+class GroupByGenerator(object):
+  """ A custom generator used by groupby to allow for len() calls on the
+      generator returned.
+  """
+
+  __slots__ = ['iterable', 'beginning', 'length', 'end', 'index']
+
+  def __init__(self, iterable, beginning, length):
+    self.iterable = iterable
+    self.beginning = beginning
+    self.length = length
+    self.end = self.beginning + self.length
+    self.index = 0
+
+
+  def __iter__(self):
+    for i in xrange(self.beginning, self.end):
+      self.index = i
```

Comment:
This loop shouldn't change the object, otherwise there will be a bug that you can't iterate over a `groupby` result in two places at once. (Not necessarily in parallel)

We don't need to worry about the `__len__` result updating as you iterate. It should just return the length of the original list subset.